Update main.go, remove XML tags in the GDScript docstrings

NathanLovato opened the pull request:
This PR modifies the doc-comments to allow us to generate a code reference for nakama-godot, using [gdscript docs maker](https://github.com/GDQuest/gdscript-docs-maker).

It changes the doc-comment style from C#-like to GDScript-like, i.e. a plain `# My doc comment`.

Note that I used the latest `apigrpc.swagger.json` file from http://github.com/heroiclabs/nakama to regenerate the GDScript API, so there is one new Facebook-related feature in this commit. Also, there is currently no processing done on a handful of XML tags in the source project, i.e. 

```xml
# Send data to the server with an asynchronous operation.
# <param name="p_buffer">The buffer with the message to send.</param>
# <param name="p_reliable">If the message should be sent reliably (will be ignored by some protocols).</param>
```

This is in part because we don't have a standard yet for GDScript to document parameters. It might be JS-style in the end, i.e. `@param p_buffer - The buffer with the message to send`, or something else.

mofirouz commented:
Just chiming to say this is an absolutely epic addition - thank you for contributing to the community ❤️ 

I'll leave the full review to @Faless and @novabyte.

NathanLovato commented:
Thanks. We already talked with @Faless on IRC beforehand about that.

Maybe, how would you like me to go about replacing the XML tags copied from the source JSON data? With `sed`, after running main.go, or should I use go templates directly? That or anything else is fine by me.

novabyte commented:
@NathanLovato can you add more detail for me to what you mean by:

> copied from the source JSON data?

I'm a little stuck on the intent of the remainder of the change.

NathanLovato commented:
> can you add more detail for me to what you mean by:
> > copied from the source JSON data?

Some docstrings generated by codegen.go have XML built into them, like `<param></param>` tags.

The thing is, GDScript docstrings have no markup, like Python docstrings.  The Godot language server will treat that XML as regular text. This will display as-is in a code reference generated from your API, or with inline docs from the Godot language server.

E.g. if you code in VSCode or Atom with the Godot plugin, and type `NakamaHTTPAdapter.send_async`, you will get a popup that shows this:

```
Send a HTTP request.
<param name="method">HTTP method to use for this request.</param>
<param name="uri">The fully qualified URI to use.</param>
<param name="headers">Request headers to set.</param>
<param name="body">Request content body to set.</param>
<param name="timeoutSec">Request timeout.</param>
<returns>A task which resolves to the contents of the response.</returns>
````

So I'd need to use regex to make the docstrings work for GDScript, something like:

```
@method - HTTP method to use for this request.
```

I wrote that quickly, but looking at the golang ref, it should be pretty quick in go. I'll give it a go.

NathanLovato commented:
Done, I removed all remaining XML tags. Now the project should be ready to generate an online code reference with https://github.com/GDQuest/gdscript-docs-maker

Faless commented on addons/com.heroiclabs.nakama/api/NakamaAPI.gd:
```diff
@@ -3376,9 +2907,36 @@ class ApiClient extends Reference:
-	### Authenticate a user with Apple's GameCenter against the server.
-	### </summary>
+	# Authenticate a user with a Facebook Instant Game token against the server.
+	func authenticate_facebook_instant_game_async(
```
This will need the `NakamaClient.gd` counterpart (e.g.):
```gdscript
# Authenticate a user with a Facebook Instant Game token against the server.
# @param p_signed_player_info - Facebook Instant Game signed info from Facebook SDK.
# @param p_username - A username used to create the user. May be `null`.
# @param p_create - If the user should be created when authenticated.
# @param p_import - If the Facebook friends should be imported.
# @param p_vars - Extra information that will be bundled in the session token.
# A task which resolves to a session object.
func authenticate_facebook_instant_game_async(p_signed_player_info : String, p_username = null, p_create : bool = true, p_import : bool = true, p_vars = null) -> NakamaSession:
        return _parse_auth(yield(_api_client.authenticate_facebook_instant_game_async(server_key, "",
                NakamaAPI.ApiAccountFacebookInstantGame.create(NakamaAPI, {
                        "signed_player_info": p_signed_player_info,
                        "vars": p_vars
                }), p_create, p_username, p_import), "completed"))
```

Same goes for link/unlink.

NathanLovato commented:
I addressed your review comments and tried to add the missing facebook method. Note that I haven't tested the functions as I don't have a Godot project set up for this right now, I'm just diving into the API.

<hr>

How about using the GDScript code formatter to auto-format everything? There's one that follows the official GDScript style guide, and that works great: https://github.com/Scony/godot-gdscript-toolkit/.

The current layout, with long lines, makes it hard to work with two documents side by side, or to read the source code, even with soft wrapping.

Faless commented:
IIRC `p_username : String = null` is not valid yet in GDScript (typed parameters cannot be nulled).
Should be `p_username = null` without type hinting. I hope we can get that in 4.0, will have to see how the GDScript work proceed.

NathanLovato commented:
Fixed! It's good to know - in the team we avoid using `null` so I hadn't come across that. I'm looking forward to the type system becoming a bit more mature, there are still quite a few gotchas right now.

novabyte commented:
@Faless I haven't followed this discussion very well. Is this good to merge?

Faless commented:
It still needs some fix, in the Facebook part, but I can merge it and apply as a followup:
```diff
diff --git a/addons/com.heroiclabs.nakama/client/NakamaClient.gd b/addons/com.heroiclabs.nakama/client/NakamaClient.gd
index 70cb5ea..19801b7 100644
--- a/addons/com.heroiclabs.nakama/client/NakamaClient.gd
+++ b/addons/com.heroiclabs.nakama/client/NakamaClient.gd
@@ -140,12 +140,12 @@ func authenticate_facebook_async(p_token : String, p_username = null, p_create :
 # @param p_import - If the Facebook friends should be imported.
 # @param p_vars - Extra information that will be bundled in the session token.
 # Returns a task which resolves to a session object.
-func authenticate_facebook_instant_game_async(p_signed_player_info : String, p_username = null, p_create : bool = true, p_import : bool = true, p_vars = null) -> NakamaSession:
+func authenticate_facebook_instant_game_async(p_signed_player_info : String, p_username = null, p_create : bool = true, p_vars = null) -> NakamaSession:
 		return _parse_auth(yield(_api_client.authenticate_facebook_instant_game_async(server_key, "",
 				NakamaAPI.ApiAccountFacebookInstantGame.create(NakamaAPI, {
 						"signed_player_info": p_signed_player_info,
 						"vars": p_vars
-				}), p_create, p_username, p_import), "completed"))
+				}), p_create, p_username), "completed"))
 
 # Authenticate a user with Apple Game Center.
 # @param p_bundle_id - The bundle id of the Game Center application.
@@ -372,12 +372,12 @@ func link_facebook_async(p_session : NakamaSession, p_token : String) -> NakamaA
 # @param p_token - An OAuth access token from the Facebook SDK.
 # @param p_import - If the Facebook friends should be imported.
 # Returns a task which represents the asynchronous operation.
-func link_facebook_instant_game_async(p_session : NakamaSession, p_token : String) -> NakamaAsyncResult:
+func link_facebook_instant_game_async(p_session : NakamaSession, p_signed_player_info : String) -> NakamaAsyncResult:
 	return _api_client.link_facebook_instant_game_async(
 		p_session.token,
-		NakamaAPI.ApiAccountFacebook.create(
+		NakamaAPI.ApiAccountFacebookInstantGame.create(
 			NakamaAPI, {
-			"token": p_token
+				"signed_player_info": p_signed_player_info
 			})
 		)
 
@@ -657,11 +657,13 @@ func unlink_facebook_async(p_session : NakamaSession, p_token : String) -> Nakam
 # @param p_session - The session of the user.
 # @param p_token - An OAuth access token from the Facebook SDK.
 # Returns a task which represents the asynchronous operation.
-func unlink_facebook_instant_game_async(p_session : NakamaSession, p_token : String) -> NakamaAsyncResult:
+func unlink_facebook_instant_game_async(p_session : NakamaSession, p_signed_player_info : String) -> NakamaAsyncResult:
 	return _api_client.unlink_facebook_instant_game_async(
 		p_session.token,
-		NakamaAPI.ApiAccountFacebook.create(
-			NakamaAPI, {"token": p_token}))
+		NakamaAPI.ApiAccountFacebookInstantGame.create(NakamaAPI, {
+			"signed_player_info": p_signed_player_info
+		})
+	)
 
 # Unlink a Game Center profile from the user account owned by the session.
 # @param p_session - The session of the user.
```

NathanLovato commented:
Thanks!

Faless commented:
> How about using the GDScript code formatter to auto-format everything? There's one that follows the official GDScript style guide, and that works great: https://github.com/Scony/godot-gdscript-toolkit/.

Sorry, I forgot to answer to this, I think it's an excellent idea. maybe after #29 is merged to avoid a dreadful rebase ;-)
